### PR TITLE
chore: assessor calcTokenPrice and tests

### PR DESCRIPTION
**Pull request by manuelpolzhofer**
- added tests for assessor
- handle edge cases for calcTokenPrice
- simpler tranche mock

### CODE REVIEW

File: src/core/lender/test/mock/tranche.sol [hunk 1]
```diff
@@ -12,43 +11,24 @@
...
-    uint public rate;
-    uint public reserve;
-    int public loanAmount;
-
+    mapping (bytes32 => uint) calls;
+    mapping (bytes32 => uint) returnValues;
+    function setReturn(bytes32 name, uint returnValue) public {
+        returnValues[name] = returnValue;
+    }
+    function funcBody(bytes32 name) internal returns (uint) {
```
Comment (charleenfei) [inline src/core/lender/test/mock/tranche.sol]: 👍 this is cool, we should introduce this pattern for all of our mocks

File: src/core/lender/assessor.sol [hunk 2]
```diff
@@ -58,19 +61,25 @@ contract Assessor is DSNote,DSMath {
         else revert();
     }
 
+    function calcAssetValue(address tranche) public returns(uint) {
+        uint trancheReserve = TrancheLike(tranche).balance();
+        uint poolValue = pile.debt();
+        if (tranche == junior) {
+            return calcJuniorAssetValue(poolValue, trancheReserve, seniorDebt());
+        }
+        return calcSeniorAssetValue(poolValue, trancheReserve, SeniorTrancheLike(tranche).debt(), juniorReserve());
```
Comment (charleenfei) [inline src/core/lender/assessor.sol]: you could just used seniorDebt() instead of SeniorTrancheLike(tranche).debt()

File: src/core/lender/test/mock/tranche.sol (hunk 1 above)
Comment (ilinzweilin) [reply src/core/lender/test/mock/tranche.sol]: sounds good. How about calling the function simply -> call()

File: src/core/lender/assessor.sol [hunk 3]
```diff
@@ -58,19 +61,25 @@ contract Assessor is DSNote,DSMath {
...
-        } else {
-            assetValue =  calcSeniorAssetValue(poolValue, trancheReserve, SeniorTrancheLike(tranche_).debt(), juniorReserve());
+        uint tokenSupply = TrancheLike(msg.sender).tokenSupply();
+        uint assetValue = calcAssetValue(msg.sender);
+        if (tokenSupply == 0) {
+            return ONE;
         }
-
-        return rdiv(assetValue, TrancheLike(tranche_).tokenSupply());
+        if (assetValue == 0) {
```
[the remark this answers is not in the document]
Comment (manuelpolzhofer) [reply src/core/lender/assessor.sol]: if `tokenSupply == 0` it is okay for `assetValue == 0`  because it is the inital setup

File: src/core/lender/test/assessor.t.sol [hunk 4]
```diff
@@ -1,102 +1,126 @@
...
+        tokenSupply = 2.7182818284590452 ether;
+        debt = 3.14159265359 ether;
+        senior.setReturn("debt", debt);
+        pile.setDebtReturn(debt);
+        assetValue = assessor.calcAssetValue(address(senior));
+        // sanity check
+        assertEq(assetValue, debt);
+        senior.setReturn("tokenSupply",tokenSupply);
+        tokenPrice = senior.doCalcTokenPrice(assessor_);
+        assertEq(tokenPrice, rdiv(assetValue, tokenSupply));
```
[the remark this answers is not in the document]
Comment (manuelpolzhofer) [reply src/core/lender/test/assessor.t.sol]: yes. also for token price with junior tranche

File: src/core/lender/assessor.sol (hunk 2 above)
Comment (manuelpolzhofer) [reply src/core/lender/assessor.sol]: we could but I think it is a bit cheaper not doing it

File: src/core/lender/assessor.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (manuelpolzhofer) [reply src/core/lender/assessor.sol]: if `tokenSupply == 0 ` we can always return `ONE`
